nemo/collections/tts/models/fastspeech2.py

```diff
@@ -0,0 +1,244 @@
+# limitations under the License.
+
+import json
+import re
+from dataclasses import dataclass
+from typing import Any, Dict, Optional
+
+import torch
+from hydra.utils import instantiate
+from omegaconf import MISSING, DictConfig, OmegaConf, open_dict
+from pytorch_lightning.loggers import LoggerCollection, TensorBoardLogger
+
+from nemo.collections.asr.parts import parsers
+from nemo.collections.tts.helpers.helpers import plot_spectrogram_to_numpy
+from nemo.collections.tts.losses.fastspeech2loss import DurationLoss, L2MelLoss
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.core.classes.common import typecheck
+from nemo.core.neural_types.elements import *
+from nemo.core.neural_types.neural_type import NeuralType
+from nemo.utils import logging
+
+
+@dataclass
+class FastSpeech2Config:
+    preprocessor: Dict[Any, Any] = MISSING
+    encoder: Dict[Any, Any] = MISSING
+    decoder: Dict[Any, Any] = MISSING
+    variance_adaptor: Dict[Any, Any] = MISSING
+    train_ds: Optional[Dict[Any, Any]] = None
+    validation_ds: Optional[Dict[Any, Any]] = None
+
+
+class FastSpeech2Model(SpectrogramGenerator):
+    """FastSpeech 2 model used to convert between text (phonemes) and mel-spectrograms."""
+
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        if isinstance(cfg, dict):
+            cfg = OmegaConf.create(cfg)
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        schema = OmegaConf.structured(FastSpeech2Config)
+        # ModelPT ensures that cfg is a DictConfig, but do this second check in case ModelPT changes
+        if isinstance(cfg, dict):
+            cfg = OmegaConf.create(cfg)
+        elif not isinstance(cfg, DictConfig):
+            raise ValueError(f"cfg was type: {type(cfg)}. Expected either a dict or a DictConfig")
+        # Ensure passed cfg is compliant with schema
+        OmegaConf.merge(cfg, schema)
+
+        self.pitch = cfg.add_pitch_predictor
+        self.energy = cfg.add_energy_predictor
+        self.duration = True
+        self.duration_coeff = cfg.duration_coeff
+
+        self.audio_to_melspec_preprocessor = instantiate(self._cfg.preprocessor)
+        self.encoder = instantiate(self._cfg.encoder)
+        self.mel_decoder = instantiate(self._cfg.decoder)
+        self.variance_adapter = instantiate(self._cfg.variance_adaptor)
+        self.loss = L2MelLoss()
+        self.mseloss = torch.nn.MSELoss()
+        self.durationloss = DurationLoss()
+
+        self.log_train_images = False
+
+        # Parser and mappings are used for inference only.
+        self.parser = parsers.make_parser(name='en')
+        with open(cfg.mappings_filepath, 'r') as f:
+            mappings = json.load(f)
+            self.word2phones = mappings['word2phones']
+            self.phone2idx = mappings['phone2idx']
+
+    @typecheck(
+        input_types={
+            "spec_len": NeuralType(('B'), LengthsType(), optional=True),
```

Comment:
Is this input optional?

Reply:
I've made it optional so that we can use `generate_spectrogram()`, which doesn't have a ground-truth spectrogram and therefore no spec_len to pass in.

Reply:
Should we default it to None then?

Reply:
Ah yep, good catch.